### ISSUE TITLE
Server DataSource example issues with ValueBackendDataSource.read when client has default subscription

### ISSUE THREAD

**Issue by piperoc** [open62541pp/open62541pp#281]
The use case is possibly very common.

I cloned and built the code with the example. 

I then execute the `server_datasource` example, and connect to it with a generic client (in these screenshots I used UAExpert, but I have the same behavior with Prosys).

![datasource1](https://github.com/open62541pp/open62541pp/assets/103532925/6af9c98f-deab-48af-a8fe-23a742bff322)

Here you can see the normal subscription window. But the data will not update in the data view. I can perform some ad hoc _reads_ and will see the value changes, but not in the monitored item.

Therefore, I change the monitored item settings to trigger the change against the timestamp like so:
![dATAsource2](https://github.com/open62541pp/open62541pp/assets/103532925/f04e845b-9562-4d8b-8be3-8dc4d04ae8b6)

(That's probably something that a novice user won't know right away).

Tat way I can see the monitored item value changing as expected.

So, I tried building the `open62541` examples since I remember them working fine. I started the `tutorial_server_datasource` and everything works right away on the client side without needing to change the Monitored Item Settings.

Because of that, I made this small change in the `open62541pp/examples/server_datasource.cpp` file for the read callback:

```cpp
...
// Define data source with its read and write callbacks
    opcua::ValueBackendDataSource dataSource;
    dataSource.read = [&](opcua::DataValue& value, const opcua::NumericRange&, bool timestamp) {
        // Increment counter before every read
        counter++;
        value.getValue().setScalarCopy(counter);
        if (timestamp) {
            value.setSourceTimestamp(opcua::DateTime::now());
        }
        std::cout << "Read counter from data source: " << counter << "\n";
        return UA_STATUSCODE_GOOD;
    };
...
```
Basically, instead of `value.getValue().setScalar(counter);` I have `value.getValue().setScalarCopy(counter);` which copies the value. 

I assume this would make a better example to someone who is trying the code and would not understand the details of client side subscription filtering.

Also, I would like a good explanation on why the copy would trigger the value callback and the no-copy option would not. With that, what would be a good use case and motivation to choose the no-copy option (beside the obvious memory save).

Thanks again, I hope this helps others.

**Comment by lukasberbuer** [comment]
Thank you @piperoc for this helpful analysis! This issue was also discovered by @jh-isw but I couldn't figure out the reason.
Changing `setScalar` to `setScalarCopy` indeed fixes this. The reason is simple:

open62541 checks internally if the value of the monitored item was changed within [detectValueChange](https://github.com/open62541/open62541/blob/v1.3.10/src/server/ua_subscription_datachange.c#L77-L126). The last datavalue is stored in the monitored item. BUT if the variant only points to a variable (`setScalar`), changing the variable value will update both the current and the last value and therefore won't trigger a data change notification.

If you change the monitored item settings in your client to trigger data change notifications also for timestamp changes, it will work with `setScalar` as well:
![image](https://github.com/open62541pp/open62541pp/assets/36054362/86c7efb6-98eb-4f58-bc1e-b97f63e9ec09)
